source/simulator/stokes_matrix_free.cc

```diff
@@ -1222,12 +1290,21 @@ namespace aspect
   template <int dim, int velocity_degree>
   void StokesMatrixFreeHandlerImplementation<dim, velocity_degree>::evaluate_material_model ()
   {
-    // Create active viscosity vector
     dealii::LinearAlgebra::distributed::Vector<double> active_viscosity_vector(dof_handler_projection.locally_owned_dofs(),
                                                                                sim.triangulation.get_communicator());
-    {
-      const QGauss<dim> quadrature_formula (sim.parameters.stokes_velocity_degree+1);
 
+    const QGauss<dim> quadrature_formula (sim.parameters.stokes_velocity_degree+1);
+    FEValues<dim> fe_values_projection (*(sim.mapping),
+                                        fe_projection,
+                                        quadrature_formula,
+                                        update_values);
+    std::vector<double> values_on_quad;
```

Comment:
can you move this down to where it is used?

Reply:
I end up needing these for the multilevel computations as well, which is why I need to define them here. I can move the `values_on_quad` vector and just redefine for multilevel. Should I do the same with QGauss and FEValues?

Reply:
try to move things down where they are used (these are only needed after the project* call, right?).

Reply:
I end up using `fe_values_projection` in lines 1403 and 1479 and similarly for `quadrature_formula` and `values_on_quad`. I can redefine them in the various different blocks if that looks nicer?

Reply:
I was just confused that the projection right below these lines does not use it. If you think this is still sensible this way, we can keep it as is.

Reply:
Oh, I didn't realize it wasn't used anymore.. Yes, you are right. I've moved two of the variables down to the next block.